packages/toaster/components/Toaster.tsx

```diff
@@ -15,124 +15,96 @@ interface ToasterProps {
   children?: Toast | Toast[];
 }
 
-class Toaster extends React.PureComponent<ToasterProps, { toasts: Toast[] }> {
-  public timeouts: number[] = [];
+const Toaster = ({ children }: ToasterProps) => {
+  const timeouts = React.useRef<number[]>([]);
+  const toastChildren = React.Children.toArray(children) as Toast[];
 
-  constructor(props) {
-    super(props);
+  const [toasts, setToasts] = React.useState<Toast[]>(toastChildren);
 
-    this.cloneToast = this.cloneToast.bind(this);
-    this.dismissToast = this.dismissToast.bind(this);
-    this.clearTimeouts = this.clearTimeouts.bind(this);
-    this.restartTimeouts = this.restartTimeouts.bind(this);
-    this.setTimer = this.setTimer.bind(this);
-
-    this.state = {
-      toasts: React.Children.toArray<Toast>(this.props.children).map(
-        this.cloneToast
-      )
-    };
-  }
+  const handleDismissToast = (id: ToastId = "") => {
+    const toastToDismiss = toasts.find(toast => id === toast.props.id);
+    if (toastToDismiss && toastToDismiss.props.onDismiss) {
+      toastToDismiss.props.onDismiss();
+    }
+    setToasts(toasts.filter(toast => id !== toast.props.id));
+  };
+
+  const restartTimeouts = () => {
+    toasts.forEach((toast: Toast) => {
+      const toastKey = toast.key as number;
+      if (toast.props.autodismiss) {
+        timeouts.current.push(
+          window.setTimeout(() => {
+            handleDismissToast(toast.props.id);
+          }, DELAY_TIME + MARGINAL_DELAY * toastKey)
+        );
+      }
+    });
+  };
 
-  public componentWillUnmount() {
-    this.clearTimeouts();
-  }
+  const clearTimeouts = () => {
+    timeouts.current.forEach(clearTimeout);
+  };
 
-  public componentDidMount() {
-    this.restartTimeouts();
-  }
+  React.useEffect(() => {
+    return () => {
+      clearTimeouts();
+    };
+  }, []);
 
-  public UNSAFE_componentWillReceiveProps(props: ToasterProps) {
-    const children = React.Children.toArray(props.children);
-    const childIds = children.map(toast => toast.props.id);
-    const currentIds = this.state.toasts.map(toast => toast.props.id);
+  React.useEffect(() => {
+    const childIds = toastChildren.map(toast => toast.props.id);
+    const currentIds = toasts.map(toast => toast.props.id);
 
     if (
       !currentIds.every(e => childIds.includes(e)) ||
       !currentIds.length ||
       childIds.length !== currentIds.length
     ) {
-      this.setState(
-        () => ({ toasts: children.map(this.cloneToast) }),
-        () => {
-          this.state.toasts
-            ?.filter(t => t.props.autodismiss)
-            .map(this.setTimer);
-        }
-      );
+      setToasts(toastChildren);
     }
-  }
-
-  public render() {
-    return (
-      <div
-        className={cx(toaster, margin("all"), marginAt.medium("all", "l"))}
-        data-cy="toaster"
+  }, [children]);
+
+  React.useEffect(() => {
+    restartTimeouts();
+  }, [toasts]);
```

Comment:
effect should be dependent on `restartTimeouts` and not on `toasts`

Reply:
actually I was off here. This was called in the old componentDidMount, so I am adding it to the useEffect without any dependencies to recreate that.

Reply:
on second look the reason I did this was because this was called on mount in the class, but then again in the component will receive props. Both seemed to be based on changes to `toasts` state. So, now I am using useEffect to change this when toasts change.